File: src/main/java/com/ada/conversation/application/dto/MessageExecutionState.java
package com.ada.conversation.application.dto;

public sealed interface MessageExecutionState
    permits MessageExecutionState.Received,
        MessageExecutionState.FilteringCommand,
        MessageExecutionState.SelectingContext,
        MessageExecutionState.CreatingContext,
        MessageExecutionState.InvokingModel,
        MessageExecutionState.InvokingTool,
        MessageExecutionState.Completed,
        MessageExecutionState.Failed {
  String code();

  String detail();

  record Received() implements MessageExecutionState {
    public String code() {
      return "received";
    }

    public String detail() {
      return null;
    }
  }

  record FilteringCommand() implements MessageExecutionState {
    public String code() {
      return "filtering_command";
    }

    public String detail() {
      return null;
    }
  }

  record CreatingContext() implements MessageExecutionState {
    public String code() {
      return "creating_context";
    }

    public String detail() {
      return null;
    }
  }

  record SelectingContext() implements MessageExecutionState {
    public String code() {
      return "selecting_context";
    }

    public String detail() {
      return null;
    }
  }

  record InvokingModel(String model) implements MessageExecutionState {
    public String code() {
      return "invoking_model";
    }

    public String detail() {
      return model;
    }
  }

  record InvokingTool(String tool) implements MessageExecutionState {
    public String code() {
      return "invoking_tool";
    }

    public String detail() {
      return tool;
    }
  }

  record Completed() implements MessageExecutionState {
    public String code() {
      return "completed";
    }

    public String detail() {
      return null;
    }
  }

  record Failed(String reason) implements MessageExecutionState {
    public String code() {
      return "failed";
    }

    public String detail() {
      return reason;
    }
  }
}

File: src/main/kotlin/com/ada/conversation/application/dto/MessageExecutionState.kt
package com.ada.conversation.application.dto

sealed interface MessageExecutionState {
    val code: String
    val detail: String?

    data object Received : MessageExecutionState {
        override val code = "received"
        override val detail: String? = null
    }

    data object FilteringCommand : MessageExecutionState {
        override val code = "filtering_command"
        override val detail: String? = null
    }

    data object CreatingContext : MessageExecutionState {
        override val code = "creating_context"
        override val detail: String? = null
    }

    data class InvokingModel(val model: String) : MessageExecutionState {
        override val code = "invoking_model"
        override val detail = model
    }

    data class InvokingTool(val tool: String) : MessageExecutionState {
        override val code = "invoking_tool"
        override val detail = tool
    }

    data object Completed : MessageExecutionState {
        override val code = "completed"
        override val detail: String? = null
    }

    data class Failed(val reason: String) : MessageExecutionState {
        override val code = "failed"
        override val detail = reason
    }
}
